### PR TITLE
feat(orm): add streaming QuerySet iterator

### LINKED ISSUE #5846: Feature: Add streaming QuerySet iterator
## Prerequisites

- [x] I have searched existing issues to verify this feature hasn't been requested

## Summary

Add a genuinely bounded QuerySet iterator backed by a lifetime-bound row stream in the ORM executor contract.

Split from #5812. Target branch: `develop/0.4.0`.

## Motivation / Problem

`OrmExecutor` currently exposes `fetch_all`, so large QuerySet results must be fully materialized. Applications need row-by-row model decoding without unbounded QuerySet caching or misleading pagination-based emulation.

## Proposed Solution

- Extend the ORM executor contract with a lifetime-bound row stream capability.
- Add `iterator_with_executor()` returning a stream tied to the borrowed executor.
- Return `Result<Model>` per item so mid-stream backend and decode failures remain observable.
- Treat `chunk_size` as a driver fetch or bounded-buffer hint.
- Never call `fetch_all` and never emulate streaming with repeated `LIMIT` / `OFFSET` queries.
- Return an explicit capability error for executors without streaming support.
- Release driver resources through RAII when the stream completes, errors, is dropped early, or is cancelled.
- Ensure a `none()` QuerySet invokes no executor method.
- Add bounded-buffer instrumentation, cancellation and error tests, built-in backend integration tests, and user-facing documentation.

### Durable implementation and verification contracts

- Treat the lifetime-bound `iterator_with_executor()` path as the caller-owned-executor variant, and keep equivalent supported execution paths on the borrowed executor rather than reacquiring a pooled connection.
- Keep the streamed query represented by `reinhardt-query` statements and expressions rather than unchecked SQL strings in QuerySet code.
- Add tests for empty streams, early drop, cancellation, and mid-stream backend and decode errors.
- Instrument a mock stream to prove that the maximum number of buffered rows never exceeds the configured bound.
- Add built-in backend integration tests that prove row-by-row delivery without a QuerySet-level result cache.
- Update the relevant crate README, `lib.rs` module documentation, and ORM user-facing documentation. Include typed-field and caller-owned-executor examples where applicable, state backend limitations explicitly, and document the intentionally stricter-than-Django rejection of eager materialization and pagination-based streaming fallbacks.

## Alternatives Considered

1. Repeated `LIMIT` / `OFFSET` queries. Concurrent changes alter result consistency and large offsets degrade performance.
2. `fetch_all` followed by an iterator. This is not bounded streaming.
3. Silent fallback for unsupported executors. This hides materially different memory behavior.

## Is this a breaking change?

No - backward compatible

## Migration Path (if breaking)

Not applicable. Existing eager execution methods remain available.

## Additional Context

This issue intentionally excludes retrieval helpers, explain, date projection, row locking, and pagination-based iterator fallback.

### PR DESCRIPTION
**Pull request by kent8192**
## Summary

This PR addresses:

- Add lifetime-bound `RowStream` capabilities to database and transaction executor contracts.
- Add `QuerySet::iterator_with_executor()` / `iterator_with_db()` and `QuerySet::none()` without eager materialization or LIMIT/OFFSET pagination.
- Stream typed models from PostgreSQL, MySQL, and SQLite executors with per-item errors, explicit unsupported-capability errors, and RAII cleanup on completion, error, cancellation, or drop.
- Document caller-owned executor and typed-field streaming usage.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (code change that neither fixes a bug nor adds a feature)
- [x] Documentation update
- [x] Performance improvement
- [x] Code quality improvements
- [ ] CI/CD changes
- [ ] Other (please describe):

## Breaking Change Assessment

- [x] **No** - This change is backward compatible
- [ ] **Yes** - This change breaks existing public API or behavior

## Motivation and Context

Large querysets currently require eager row collection before model decoding. This change introduces a lifetime-bound streaming path whose resource lifetime is tied to the caller-owned executor. `chunk_size` is passed as a driver or bounded-buffer hint; the implementation never falls back to `fetch_all()` or repeated LIMIT/OFFSET queries.

Fixes #5846

Related to: #5812

## How Was This Tested?

- `cargo test --offline -p reinhardt-db --no-default-features --features orm,sqlite --test queryset_streaming -- --test-threads=1` (8 passed)
- `cargo check --offline -p reinhardt-db --no-default-features --features orm,sqlite,mysql,postgres --lib`
- `cargo clippy --offline -p reinhardt-db --no-default-features --features orm,sqlite,mysql,postgres --lib -- -D warnings`
- `RUSTDOCFLAGS='-D warnings' cargo doc --offline -p reinhardt-db --no-deps`
- `cargo fmt --all -- --check`
- `git diff --check`

The integration test suite covers bounded buffering, empty and `none()` querysets, invalid chunks, unsupported executors, cancelled polls, early drop, mid-stream backend and decode errors, and an actual SQLite backend stream.

## Performance Impact

The iterator avoids queryset-level eager materialization. Built-in SQLx backends stream rows from the driver, while `chunk_size` remains a non-semantic driver/buffer hint and does not alter query ordering or result contents.

## Checklist

- [x] I have followed the [Contributing Guidelines](../blob/main/CONTRIBUTING.md)
- [x] I have followed the [Commit Guidelines](../blob/main/instructions/COMMIT_GUIDELINE.md)
- [x] I have updated the documentation (if applicable)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have tested with all affected database backends (if applicable)
- [x] I have formatted the code with `cargo make fmt-fix`
- [ ] I have checked the code with `cargo make clippy-check`

- [ ] I use self-hosted runner for CI (Repository owner only)

## Related Issues

- Fixes #5846
- Related to #5812

### Type Label (select one)
- [ ] `bug` - Bug fix
- [x] `enhancement` - New feature or improvement
- [ ] `documentation` - Documentation update
- [ ] `performance` - Performance improvement
- [ ] `refactoring` - Code refactoring
- [ ] `code-quality` - Code quality improvements

### Additional Labels (apply alongside type label when applicable)
- [ ] `breaking-change` - Breaking change (**MUST** match "Yes" in Breaking Change Assessment above)

### Scope Label (select all that apply)
- [x] `database` - Database layer, schema, migrations
- [ ] `auth` - Authentication, authorization, sessions
- [x] `orm` - ORM layer, models, query builder
- [ ] `http` - HTTP layer, handlers, middleware
- [ ] `routing` - URL routing, path matching
- [ ] `api` - REST API, serializers, views
- [ ] `admin` - Admin interface, admin panels
- [ ] `forms` - Form handling, validation, rendering
- [ ] `graphql` - GraphQL schema, resolvers
- [ ] `websockets` - WebSocket connections, handlers
- [ ] `i18n` - Internationalization, localization
- [ ] `ci-cd` - CI/CD workflow changes

### Priority Label (for maintainers)
- [ ] `critical` - Blocks release or major functionality
- [x] `high` - Important fix or feature
- [ ] `medium` - Normal priority
- [ ] `low` - Minor fix or enhancement

---

**Additional Context:**

Custom executors that do not implement row streaming return an explicit `Unsupported` database error. There is intentionally no eager or pagination fallback.

🤖 Generated with [Codex](https://openai.com/codex)

### CODE REVIEW

File: crates/reinhardt-db/src/orm/query.rs [hunk 1]
```diff
@@ -6126,6 +6145,148 @@ where
...
+		let context = super::execution::pgvector_context_for_select(&stmt);
+		let (sql, values) =
+			Self::build_select_for_backend(&stmt, conn.backend(), conn.is_cockroachdb())?;
+		let param_samples = values
+			.iter()
+			.map(|value| value.to_sql_literal())
+			.collect::<Vec<_>>();
+		let params = super::execution::convert_values(values);
+		let mut rows = conn.fetch_stream_with_context(sql.clone(), params, chunk_size, context)?;
+		let started = Instant::now();
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: **Fixed: stream timing now starts on the first poll**\n\nThe timing guard is constructed inside the async generator immediately before it polls the backend stream, so idle time between creating and consuming an iterator is excluded.\n\nCommit: 64c269402d — \n\n🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 2]
```diff
@@ -6126,6 +6145,148 @@ where
...
+							.orm_query_error(&sql, &format!("{error:?}"))
+							.await;
+						yield Err(error);
+						return;
+					}
+				}
+			}
+			super::instrumentation::instrumentation()
+				.orm_query_end_with_params(&sql, &param_samples, started.elapsed())
+				.await;
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: **Fixed: partially consumed streams are recorded**\n\nA drop guard records the query exactly when the streaming generator completes or is dropped after polling, so , early loop exits, and cancellation remain visible to N+1 detection.\n\nCommit: 64c269402d — \n\n🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 3]
```diff
@@ -6126,6 +6145,148 @@ where
...
+		Ok(Box::pin(async_stream::stream! {
+			while let Some(row) = rows.next().await {
+				let row = match row {
+					Ok(row) => row,
+					Err(error) => {
+						super::instrumentation::instrumentation()
+							.orm_query_error(&sql, &format!("{error:?}"))
+							.await;
+						yield Err(error);
+						return;
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:6402]: **Fixed: terminal errors release backend rows before yielding**\n\nBoth iterator paths hold the backend stream in an  and clear it before yielding a terminal backend or deserialization error, releasing its cursor or connection even if the consumer retains the outer stream.\n\nCommit: 64c269402d — \n\n🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/backends/dialect/postgres.rs [hunk 4]
```diff
@@ -217,6 +223,51 @@ impl DatabaseBackend for PostgresBackend {
...
+			for param in &params {
+				query = match Self::bind_value(query, param) {
+					Ok(query) => query,
+					Err(error) => {
+						yield Err(error);
+						return;
+					}
+				};
+			}
+			let rows = query.fetch(pool.as_ref());
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/backends/dialect/postgres.rs:261]: **Fixed: built-in SQLx streams now apply the chunk bound**\n\nPostgreSQL, MySQL, and SQLite pool and transaction streams use , which bounds each buffered batch while preserving one driver query and cursor-backed streaming.\n\nCommit: 64c269402d — \n\n🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 5]
```diff
@@ -1304,10 +1311,19 @@ where
...
 			from_subquery_sql: None,
 		}
 	}
 
+	/// Returns a QuerySet that is known to contain no rows.
+	///
+	/// Execution methods can use this marker to avoid invoking an executor.
+	pub fn none(mut self) -> Self {
+		self.empty_result = true;
+		self
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:1500]: **Fixed: `none()` now short-circuits every read and write execution path**

`none()` returns empty rows, zero counts, and zero affected rows before acquiring or invoking an executor, including `all`, explicit-executor reads, counts, existence checks, and `update_fields`.

Commit: 64c269402d — `crates/reinhardt-db/src/orm/query.rs:L5943`

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: **Fixed: stream timing now starts on the first poll**

The timing guard is constructed inside the async generator immediately before it polls the backend stream, so idle time between creating and consuming an iterator is excluded.

Commit: 64c269402d — `crates/reinhardt-db/src/orm/query.rs:L6221`

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: **Fixed: partially consumed streams are recorded**

A drop guard records the query exactly when the streaming generator completes or is dropped after polling, so `take`, early loop exits, and cancellation remain visible to N+1 detection.

Commit: 64c269402d — `crates/reinhardt-db/src/orm/query.rs:L59`

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:6402]: **Fixed: terminal errors release backend rows before yielding**

Both iterator paths hold the backend stream in an `Option` and clear it before yielding a terminal backend or deserialization error, releasing its cursor or connection even if the consumer retains the outer stream.

Commit: 64c269402d — `crates/reinhardt-db/src/orm/query.rs:L6222`

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/backends/dialect/postgres.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/backends/dialect/postgres.rs:261]: **Fixed: built-in SQLx streams now apply the chunk bound**

PostgreSQL, MySQL, and SQLite pool and transaction streams use `ready_chunks(chunk_size)`, which bounds each buffered batch while preserving one driver query and cursor-backed streaming.

Commit: 64c269402d — `crates/reinhardt-db/src/backends/dialect/postgres.rs:L263`

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 6]
```diff
@@ -1304,10 +1338,19 @@ where
...
+			empty_result: false,
 			from_subquery_sql: None,
 		}
 	}
 
+	/// Returns a QuerySet that is known to contain no rows.
+	///
+	/// Execution methods can use this marker to avoid invoking an executor.
+	pub fn none(mut self) -> Self {
+		self.empty_result = true;
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:1499]: Implemented in daf17ab42f: compiling an empty queryset now adds an always-false predicate, so nested subqueries preserve none() semantics.\n\n---\n*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 7]
```diff
@@ -7508,6 +7721,9 @@ where
 			if let Some(cond) = self.build_where_condition()? {
 				stmt.cond_where(cond);
 			}
+			if self.empty_result {
+				stmt.and_where(Expr::val(false));
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:7918]: Implemented in `crates/reinhardt-db/src/orm/query.rs`: `select_related_query_with_condition()` now appends an always-false predicate for `none()` querysets, so select-related SQL remains empty in every consumer path.

---
*Responded by Codex.*

File: crates/reinhardt-db/src/orm/query.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:1499]: Implemented in `crates/reinhardt-db/src/orm/query.rs`: `build_where_condition_for_write()` now emits an always-false predicate for `none()`, covering public update and delete SQL builders as well as filtered write builders.

---
*Responded by Codex.*

File: crates/reinhardt-db/tests/queryset_streaming.rs [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+	async fn commit(self: Box<Self>) -> reinhardt_core::exception::Result<()> {
+		Ok(())
+	}
+
+	async fn rollback(self: Box<Self>) -> reinhardt_core::exception::Result<()> {
+		Ok(())
+	}
+}
+
+#[tokio::test]
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/tests/queryset_streaming.rs:244]: Updated `crates/reinhardt-db/tests/queryset_streaming.rs`: all eight streaming cases now use `#[rstest]` and explicit Arrange, Act, and Assert sections.

---
*Responded by Codex.*

File: crates/reinhardt-db/tests/queryset_streaming.rs [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+
+#[cfg(feature = "sqlite")]
+#[tokio::test]
+async fn sqlite_queryset_iterator_delivers_rows_without_query_cache() {
+	let owner = reinhardt_db::backends::DatabaseConnection::connect_sqlite("sqlite::memory:")
+		.await
+		.unwrap();
+	owner
+		.execute(
+			"CREATE TABLE streamed_articles (id INTEGER PRIMARY KEY, title TEXT NOT NULL)",
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/tests/queryset_streaming.rs]: Updated `crates/reinhardt-db/tests/queryset_streaming.rs`: the SQLite fixture now builds both schema and inserts with `reinhardt-query` builders instead of literal SQL.

---
*Responded by Codex.*

File: crates/reinhardt-db/src/orm/query.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:1499]: Implemented in `crates/reinhardt-db/src/orm/query.rs`: both composite-key lookup entry points return the normal not-found result before resolving or invoking an executor for `none()` querysets.

---
*Responded by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 10]
```diff
@@ -6365,6 +6569,9 @@ where
 	where
 		E: OrmExecutor,
 	{
+		if self.empty_result {
+			return Ok(0);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:6749]: Implemented in `crates/reinhardt-db/src/orm/query.rs`: `count()`, `exists()`, `first()`, and `get()` now short-circuit `none()` before acquiring the global connection.

---
*Responded by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 11]
```diff
@@ -26,13 +27,45 @@ use serde::{Deserialize, Serialize};
...
+	}
+}
+
+impl Drop for StreamQueryAccounting {
+	fn drop(&mut self) {
+		super::instrumentation::instrumentation().orm_query_end_with_params_sync(
+			&self.sql,
+			&self.params,
+			self.started.elapsed(),
+		);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:110]: Implemented in `crates/reinhardt-db/src/orm/query.rs`: streaming accounting now accumulates only the duration of each backend row-stream poll, excluding consumer processing and idle time between polls.

---
*Responded by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 12]
```diff
@@ -6126,6 +6199,157 @@ where
...
+				let Some(row) = row else {
+					break;
+				};
+				let row = match row {
+					Ok(row) => row,
+					Err(error) => {
+						super::instrumentation::instrumentation()
+							.orm_query_error(&sql, &format!("{error:?}"))
+							.await;
+						yield Err(error);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:6401]: Fixed in f5475311dd. `crates/reinhardt-db/src/orm/query.rs` now drops the `RowStream` before yielding a terminal backend or deserialization error, so retaining the outer model stream cannot retain a backend cursor or pooled connection. The streaming tests now assert that release immediately after each terminal error.\n\n🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 13]
```diff
@@ -6126,6 +6199,157 @@ where
...
+		let params = super::execution::convert_values(values);
+		let rows = conn.fetch_stream_with_context(sql.clone(), params, chunk_size, context)?;
+
+		Ok(Box::pin(async_stream::stream! {
+			let mut accounting = StreamQueryAccounting::new(sql.clone(), param_samples);
+			let mut rows = Some(rows);
+			loop {
+				let poll_started_at = Instant::now();
+				let row = rows.as_mut().expect("row stream is available").next().await;
+				accounting.record_poll(poll_started_at.elapsed());
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Fixed in f5475311dd. `crates/reinhardt-db/src/orm/query.rs` now times individual ready backend-stream polls; pending polls do not retain elapsed state, so cancellation resets the in-flight interval instead of charging later consumer idle time to query duration.\n\n🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 14]
```diff
@@ -26,13 +27,79 @@ use serde::{Deserialize, Serialize};
...
+
+impl Stream for TimedRowStream<'_, '_> {
+	type Item = reinhardt_core::exception::Result<crate::backends::types::Row>;
+
+	fn poll_next(mut self: Pin<&mut Self>, context: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+		let started_at = Instant::now();
+		let result = self.rows.as_mut().poll_next(context);
+		if result.is_ready() {
+			self.accounting.record_poll(started_at.elapsed());
+		}
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:213]: Implemented in 6315cadf73520228ba809822d4d76368ec67d860: `TimedRowStream` now retains the start time across `Pending` polls and records the elapsed wait when the backend yields a ready result.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 15]
```diff
@@ -26,13 +27,79 @@ use serde::{Deserialize, Serialize};
...
+	}
+}
+
+impl Drop for StreamQueryAccounting {
+	fn drop(&mut self) {
+		super::instrumentation::instrumentation().orm_query_end_with_params_sync(
+			&self.sql,
+			&self.params,
+			self.duration,
+		);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:110]: Implemented in 6315cadf73520228ba809822d4d76368ec67d860: streaming accounting is disarmed after backend or deserialization errors, so failed streams report the error without recording a completed query.

---
*Generated by Codex.*

File: crates/reinhardt-db/tests/queryset_streaming.rs [hunk 16]
```diff
@@ -0,0 +1,476 @@
...
+	{
+		Ok(_) => panic!("unsupported executor must be rejected"),
+		Err(error) => error,
+	};
+	assert_eq!(error.database_kind(), Some(DatabaseErrorKind::Unsupported));
+}
+
+#[cfg(feature = "sqlite")]
+#[rstest]
+#[tokio::test]
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/tests/queryset_streaming.rs:452]: Implemented in 6315cadf73520228ba809822d4d76368ec67d860: the registry-backed SQLite iterator test now uses the named `queryset_streaming_connection_registry` serial group.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 17]
```diff
@@ -26,13 +27,100 @@ use serde::{Deserialize, Serialize};
...
+
+impl Stream for TimedRowStream<'_, '_> {
+	type Item = reinhardt_core::exception::Result<crate::backends::types::Row>;
+
+	fn poll_next(mut self: Pin<&mut Self>, context: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+		if self.pending_since.is_none() {
+			self.pending_since = Some(Instant::now());
+		}
+		let result = self.rows.as_mut().poll_next(context);
+		if result.is_ready() {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented in 8e483c2c9f: polling time is now accounted per backend poll, so a cancelled Pending next() cannot carry elapsed time into a later poll.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 18]
```diff
@@ -6126,6 +6250,159 @@ where
...
+					Err(error) => {
+						let error = Error::from(DatabaseError::new(
+							DatabaseErrorKind::Serialization,
+							format!("Deserialization error: {error}"),
+						));
+						super::instrumentation::instrumentation()
+							.orm_query_error(&sql, &format!("{error:?}"))
+							.await;
+						drop(rows.take());
+						accounting.disarm_completion();
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented in 8e483c2c9f: model decoding failures now preserve completion accounting, while backend streaming failures remain disarmed.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 19]
```diff
@@ -6126,6 +6236,157 @@ where
...
+					break;
+				};
+				let row = match row {
+					Ok(row) => row,
+					Err(error) => {
+						super::instrumentation::instrumentation()
+							.orm_query_error(&sql, &format!("{error:?}"))
+							.await;
+						drop(rows.take());
+						yield Err(error);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:6401]: Implemented in e2aea886e4: backend-stream failures now disarm `StreamQueryAccounting` before yielding the error, preventing a false successful-completion record.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 20]
```diff
@@ -9137,6 +9436,49 @@ mod tests {
 		assert_eq!(executor.contexts, vec![Some(distance_and_vector_context())]);
 	}
 
+	#[cfg(feature = "pgvector")]
+	#[tokio::test]
+	async fn none_short_circuits_read_and_write_execution_paths() {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:9680]: Implemented in e2aea886e4: converted the affected unit test to `#[rstest]` and structured it with Arrange/Act/Assert phases.

---
*Generated by Codex.*

File: crates/reinhardt-db/README.md [hunk 21]
```diff
@@ -66,6 +66,39 @@ This crate provides the following modules:
...
+driver resources when it completes, fails, is cancelled, or is dropped early.
+
+```rust
+use futures::StreamExt;
+use reinhardt_db::orm::{Model, OrmExecutor, QuerySet};
+
+async fn stream_models<M, E>(connection: &mut E) -> reinhardt_core::exception::Result<()>
+where
+    M: Model,
+    E: OrmExecutor,
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/README.md:84]: Implemented in e2aea886e4: the streaming example now requires `DeserializeOwned`, matching `iterator_with_db`.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm.rs [hunk 22]
```diff
@@ -12,6 +12,20 @@
 //! records the SQL joins required by the filter, so application code does not
 //! need raw join builders for common FK, reverse, or M2M lookups.
 //!
+//! ## Streaming QuerySets
+//!
+//! [`QuerySet::iterator_with_db`](crate::orm::QuerySet::iterator_with_db) and
+//! [`QuerySet::iterator_with_executor`](crate::orm::QuerySet::iterator_with_executor)
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm.rs]: Implemented in e2aea886e4: removed redundant explicit targets from the streaming API intra-doc links.

---
*Generated by Codex.*

File: crates/reinhardt-db/src/orm/query.rs [hunk 23]
```diff
@@ -26,13 +27,86 @@ use serde::{Deserialize, Serialize};
...
+		Self { rows, accounting }
+	}
+}
+
+impl Stream for TimedRowStream<'_, '_> {
+	type Item = reinhardt_core::exception::Result<crate::backends::types::Row>;
+
+	fn poll_next(mut self: Pin<&mut Self>, context: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+		let started_at = Instant::now();
+		let result = self.rows.as_mut().poll_next(context);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented in 605adf9452: `RowStreamWake` records the pending-to-wake interval, while the next poll discards any interval that did not wake. This includes actual database I/O without carrying cancelled-consumer idle time into a later poll.

Regression coverage: `crates/reinhardt-db/src/orm/query.rs:L120` and `crates/reinhardt-db/src/orm/query.rs:L8992`

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 24]
```diff
@@ -6126,6 +6236,158 @@ where
...
+					Err(error) => {
+						let error = Error::from(DatabaseError::new(
+							DatabaseErrorKind::Serialization,
+							format!("Deserialization error: {error}"),
+						));
+						super::instrumentation::instrumentation()
+							.orm_query_error(&sql, &format!("{error:?}"))
+							.await;
+						drop(rows.take());
+						accounting.disarm_completion();
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs]: Implemented in 605adf9452: the `iterator_with_executor` decode-error path releases the row stream but keeps completion accounting armed. Only backend stream errors disarm it, so the completed SQL execution remains visible to the N+1 scope.

Regression coverage: `crates/reinhardt-db/tests/queryset_streaming.rs:L364`

🤖 Generated with [Codex](https://openai.com/codex)

File: crates/reinhardt-db/src/orm/query.rs [hunk 25]
```diff
@@ -7508,6 +7805,9 @@ where
 			if let Some(cond) = self.build_where_condition()? {
 				stmt.cond_where(cond);
 			}
+			if self.empty_result {
+				stmt.and_where(Expr::val(false));
```
[the remark this answers is not in the document]
Comment (kent8192) [reply crates/reinhardt-db/src/orm/query.rs:7918]: Implemented in 605adf9452: `none()` now forces `LIMIT 0` in SQL rendering, including aggregate projections, so `EXISTS`, `IN`, annotation, and derived-table callers receive zero rows instead of an aggregate's synthetic count row.

Regression coverage: `crates/reinhardt-db/src/orm/query.rs:L7972` and `crates/reinhardt-db/src/orm/query.rs:L9725`

🤖 Generated with [Codex](https://openai.com/codex)
